pyinteraph/path_analysis.py

```diff
@@ -0,0 +1,473 @@
+            if graph.has_edge(node1, node2):
+                graph[node1][node2]["weight"] += 1
+            # Add edge otherwise
+            else:
+                graph.add_edge(node1, node2, weight = 1)
+    return graph
+
+
+def get_common_edges(graph, threshold):
+    """Takes in a graph where the edge weights are the count of the edge
+    and returns a list of edges which are larger than the provided 
+    threshold.
+    """
+
+    # Initialize arrays of edges and their count
+    edges = np.array(graph.edges)
+    counts = np.array([graph[e[0]][e[1]]["weight"] for e in edges])
+    # Calculate percentage occurance for each path
+    perc = counts/counts.sum()
+    # Choose paths larger than the threshold
+    common_edges = edges[perc > threshold]
+    # Convert array to list of tuples
+    common_edges = list(map(tuple, common_edges))
+    return common_edges
+
+def get_metapath(graph, res_id, res_space, node_threshold, edge_threshold):
+    """Takes in a list of paths, an edge threshold and a node threshold
+    and returns a list of metapaths where each metapath contains all 
+    nodes and edges above their respective thresholds.
+    """
+    
+    # Calculate all shortest paths
+    paths = get_all_shortest_paths(graph, res_id, res_space)
+    # Create graph from path list
+    paths_graph = get_graph_from_paths(paths)
+
+    # Get common nodes
+    common_nodes = get_common_nodes(paths, node_threshold)
+    # Get common edges
+    common_edges = get_common_edges(paths_graph, edge_threshold)
+
+    # Warn if no common nodes found
+    if len(common_nodes) == 0:
+        warn_str = f"No Nodes found with a frequency higher than: {node_threshold}"
+        log.warning(warn_str)
+    # Warn if no common edges found
+    if len(common_edges) == 0:
+        warn_str = f"No Edges found with a frequency higher than: {edge_threshold}"
+        log.warning(warn_str)
+
+    common_paths = []
+    # Common edges and nodes exist
+    if len(common_nodes) != 0 and len(common_edges) != 0:
+        # Find which paths have the common nodes and edges
+        for p in paths:
+            edges = [(p[i], p[i+1]) for i in range(len(p) - 1)]
+            # Check if required nodes are in path
+            for c_node in common_nodes:
+                if c_node in p:
+                    # Check if required edges are in path
+                    for c_edge in common_edges:
+                        # Add path if not already added
+                        if c_edge in edges and p not in common_paths:
+                            common_paths.append(p)
+        # Warn if no metapaths found
+        if len(common_paths) == 0:
+            log.warning("No metapaths found.")
+    return common_paths, paths_graph
+
+def write_table(fname, table):
+    """Save sorted table as txt file. """
+    with open(f"{fname}.txt", "w") as f:
+        for p, s, t, l, sum_w, avg_w in table:
+            f.write(f"{p}\t{s}\t{t}\t{l}\t{sum_w}\t{avg_w}\n")
+
+
+def main():
```

Comment:
leave some space between the main() and the block of code after, and in general leave space between these "title" comment lines and the code after them.